File: w1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

string get_card_number()
{
    string s;

    do
    {
        s = get_string("Number: ");
    }
    while (strlen(s) <= 0);

    return s;
}

int chr2int(char chr)
{
    return chr - '0';
}

bool valid_card(string cn)
{
  int size = strlen(cn);
  bool valid = false;

  int total = 0;

  for (int i = size - 1; i >= 0; i -= 2) {
    int digit = chr2int(cn[i]);
    total += digit;
  }

  for (int i = size - 2; i >= 0; i -= 2) {
    int digit = chr2int(cn[i]);

    digit *= 2;
    if (digit > 9) digit -= 9;

    total += digit;
  }

  return total % 10 == 0;
}

void print_card(string cn)
{
    char c[4];
    int size = strlen(cn);
    sprintf(c, "%c%c", cn[0], cn[1]);

    int fc = atoi(c);

    if ((fc == 34 || fc == 37) && size == 15)
    {
        printf("AMEX\n");
    }
    else if ((fc >= 40 && fc <= 49) && (size == 13 || size == 16))
    {
        printf("VISA\n");
    }
    else if ((fc >= 51 && fc <= 55) && size == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int main()
{
    string cn;
    // string cn = get_card_number();

    cn = "4003600000000014"; // Example VISA (20)

    // cn = "378282246310005"; // AMEX
    // cn = "371449635398431"; // AMEX
    // cn = "4222222222222"; // VISA
    // cn = "1234567890"; // INVALID (a)
    // cn = "4062901840"; // INVALID (l)
    // cn = "4222222222223"; // INVALID (c)
    // cn = "430000000000000"; // VISA
    
    bool is_valid = valid_card(cn);

    if (is_valid)
        print_card(cn);
    else
        printf("INVALID\n");
}

File: w6/sentimental-credit/credit.py
def identify_card(card_number: str):
    """
    Identify card type from nunber

    >>> identify_card("378282246310005")
    'AMEX'

    >>> identify_card("4003600000000014")
    'VISA'

    >>> identify_card("5555555555554444")
    'MASTERCARD'

    >>> identify_card("1234567890")
    """

    two_nunbers = int(card_number[:2])
    card_size = len(card_number)
    
    if two_nunbers in [34, 37] and card_size == 15:
        return "AMEX"
    elif 40 <= two_nunbers <= 49 and card_size in [13, 16]:
        return "VISA"
    elif 51 <= two_nunbers <= 55 and card_size == 16:
        return "MASTERCARD"


def is_valid(card_number: str):
    """
    Solve algorithm to determine if the card is valid

    >>> is_valid("4003600000000014")
    True

    >>> is_valid("378282246310005")
    True

    >>> is_valid("5555555555554444")
    True

    >>> is_valid("1234567890")
    False
    """

    first_list = card_number[-1::-2]
    first_iter = map(int, first_list)
    first_result = sum(first_iter)

    second_list = card_number[-2::-2]

    def duplicate_unit(digit: str):
        digit = int(digit)
        digit *= 2

        if (digit > 9): digit -= 9

        return digit

    second_iter = map(duplicate_unit, second_list)
    second_result = sum(second_iter)

    total = first_result + second_result

    return total % 10 == 0


def get_string(prompt: str):
    value = ''

    while value == '':
        value = input(prompt)

    return value


def main(card_number = None):
    """
    Program to validate credit cards

    >>> main("4003600000000014")
    VISA


    >>> main("4222222222222")
    VISA

    ---

    >>> main("378282246310005")
    AMEX

    >>> main("371449635398431")
    AMEX

    ---

    >>> main("5555555555554444")
    MASTERCARD

    ---

    >>> main("1234567890")
    INVALID

    >>> main("4062901840")
    INVALID

    >>> main("4222222222223")
    INVALID

    >>> main("430000000000000")
    INVALID
    """

    if card_number is None:
        card_number = get_string("Number: ")

    card_type = identify_card(card_number)
    
    if (identify_card(card_number) is None
        or
        not is_valid(card_number)
    ):
        print("INVALID")
    else:
        print(card_type)


if __name__ == "__main__":
    main()
